### PR TITLE
Add Tables extension

### LINKED ISSUE #404: Missing Table Binding
There doesn't seem to be a way to bind to a table. Previously there was the TableAttribute in the Microsoft.Azure.WebJobs.Extensions.Storage assembly that allowed direct access to an Azure Table.

Is there another way to achieve this? When using the TableInputAttribute in the following way

```
[Function("Bands")]
 public async Task<HttpResponseData> Execute(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get")] HttpRequestData req,
            [TableInput("Requests", Connection = "AzureWebJobsStorage")] CloudTable table,
            FunctionContext executionContext)
        {
            var tableEntity = new PayloadTableEntity(req.ReadAsString());

            await table.ExecuteAsync(TableOperation.InsertOrMerge(tableEntity));
...
        }
```

This exception is thrown:

```
System.Private.CoreLib: Exception while executing function: Functions.Bands. System.Private.CoreLib: Result: 
Failure
Exception: Microsoft.Azure.Functions.Worker.Diagnostics.Exceptions.FunctionInputConverterException: Error converting 1 input parameters 
for Function 'Bands': Cannot convert input parameter 'table' to type 'Microsoft.Azure.Cosmos.Table.CloudTable' from type 'System.String'.
[2021-04-21T07:11:43.727Z]    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultModelBindingFeature.BindFunctionInput(FunctionContext context) in D:\a\1\s\src\DotNetWorker.Core\Context\Features\DefaultModelBindingFeature.cs:line 70
[2021-04-21T07:11:43.729Z]    at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 37
[2021-04-21T07:11:43.731Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\1\s\src\DotNetWorker.Core\OutputBindings\OutputBindingsMiddleware.cs:line 13
[2021-04-21T07:11:43.733Z]    at Microsoft.Azure.Functions.Worker.GrpcWorker.InvocationRequestHandlerAsync(InvocationRequest request, IFunctionsApplication application, IInvocationFeaturesFactory invocationFeaturesFactory, ObjectSerializer serializer, IOutputBindingsInfoProvider outputBindingsInfoProvider) in D:\a\1\s\src\DotNetWorker.Grpc\GrpcWorker.cs:line 169
Stack:    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultModelBindingFeature.BindFunctionInput(FunctionContext context) in 
D:\a\1\s\src\DotNetWorker.Core\Context\Features\DefaultModelBindingFeature.cs:line 70
[2021-04-21T07:11:43.736Z]    at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 37
[2021-04-21T07:11:43.737Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\1\s\src\DotNetWorker.Core\OutputBindings\OutputBindingsMiddleware.cs:line 13
[2021-04-21T07:11:43.740Z]    at Microsoft.Azure.Functions.Worker.GrpcWorker.InvocationRequestHandlerAsync(InvocationRequest request, IFunctionsApplication application, IInvocationFeaturesFactory invocationFeaturesFactory, ObjectSerializer serializer, IOutputBindingsInfoProvider outputBindingsInfoProvider) in D:\a\1\s\src\DotNetWorker.Grpc\GrpcWorker.cs:line 169
```

### PR DESCRIPTION
**Pull request by madelinegordon**
### Issue describing the changes in this PR

resolves #404

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
  * [x] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.Functions.Worker.Extensions.Abstractions;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.Storage.Blobs
```
Comment (soninaren) [inline extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs:6]: Should be `Microsoft.Azure.Functions.Worker.Extensions.Storage.Tables`

File: extensions/Worker.Extensions.Storage.Tables/src/TableOutputAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
+﻿using Microsoft.Azure.Functions.Worker.Extensions.Abstractions;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.Storage.Blobs
```
Comment (soninaren) [inline extensions/Worker.Extensions.Storage.Tables/src/TableOutputAttribute.cs:3]: same here

**Review by soninaren — CHANGES_REQUESTED** [review]

**Review by soninaren — APPROVED** [review]
Looks good to me. Were you able to test it end to end?

File: extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.Functions.Worker.Extensions.Abstractions;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.Storage.Tables
+{
+    public class TableInputAttribute : InputBindingAttribute
+    {
+        /// <summary>Initializes a new instance of the <see cref="TableAttribute"/> class.</summary>
```
Comment (kshyju) [inline extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs:10]: nit: Did you mean "TableInputAttribute" ?

File: extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.Functions.Worker.Extensions.Abstractions;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.Storage.Tables
```
Comment (kshyju) [inline extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs:6]: Should the namespace be `Microsoft.Azure.Functions.Worker`. Check other extensions.

File: extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        /// </summary>
+        public int Take
+        {
+            get; set;
+        }
+
+        /// <summary>
+        /// Gets or sets the app setting name that contains the Azure Storage connection string.
+        /// </summary>
+        public string Connection { get; set; }
```
Comment (kshyju) [inline extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs:68]: Should this be nullable (`string?`) ?

File: extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        public string PartitionKey { get; }
+
+        /// <summary>When binding to a table entity, gets the row key of the entity.</summary>
+        /// <remarks>When binding to an entire table, returns <see langword="null"/>.</remarks>
+        public string RowKey { get; }
+
+        /// <summary>
+        /// Allow arbitrary table filter. RowKey should be null. 
+        /// </summary>
+        public string Filter
```
Comment (kshyju) [inline extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs:52]: Should this be nullable>?

File: extensions/Worker.Extensions.Storage.Tables/src/TableOutputAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.Storage.Tables
+{
+    public class TableOutputAttribute : OutputBindingAttribute
+    {
+        private readonly string _tableName;
+        private readonly string _partitionKey;
+        private readonly string _rowKey;
+
+        /// <summary>Initializes a new instance of the <see cref="TableAttribute"/> class.</summary>
```
Comment (kshyju) [inline extensions/Worker.Extensions.Storage.Tables/src/TableOutputAttribute.cs:11]: TableOutputAttribute

File: extensions/Worker.Extensions.Storage.Tables/src/Worker.Extensions.Storage.Tables.csproj [hunk 8]
```diff
@@ -0,0 +1,19 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <AssemblyName>Microsoft.Azure.Functions.Worker.Extensions.Storage.Tables</AssemblyName>
+    <RootNamespace>Microsoft.Azure.Functions.Worker.Extensions.Storage.Tables</RootNamespace>
+    <Description>Azure Table Storage extensions for .NET isolated functions</Description>
+    
+    <!--Version information-->
+    <VersionPrefix>1.0.0</VersionPrefix>
```
Comment (kshyju) [inline extensions/Worker.Extensions.Storage.Tables/src/Worker.Extensions.Storage.Tables.csproj:9]: Might be a good idea to publish this as a preview package (1.0.0-preview1), validate everything with that version and then remove the preview suffix and publish the non-preview version.

File: extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs (hunk 5 above)
Comment (madelinegordon) [reply extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs:68]: Yes you're right - for consistency I think that makes sense.

File: extensions/Worker.Extensions.Storage.Tables/src/Worker.Extensions.Storage.Tables.csproj [hunk 9]
```diff
@@ -0,0 +1,19 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <AssemblyName>Microsoft.Azure.Functions.Worker.Extensions.Storage.Tables</AssemblyName>
+    <RootNamespace>Microsoft.Azure.Functions.Worker.Extensions.Storage.Tables</RootNamespace>
+    <Description>Azure Table Storage extensions for .NET isolated functions</Description>
+    
+    <!--Version information-->
+    <VersionPrefix>1.0.0-preview1</VersionPrefix>
```
Comment (kshyju) [inline extensions/Worker.Extensions.Storage.Tables/src/Worker.Extensions.Storage.Tables.csproj:9]: nit: Can use the `VersionSuffix` element.
`<VersionSuffix>-preview1</VersionSuffix>`

File: extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        /// <remarks>When binding to a table entity, gets the name of the table containing the entity.</remarks>
+        public string TableName { get; }
+
+        /// <summary>When binding to a table entity, gets the partition key of the entity.</summary>
+        /// <remarks>When binding to an entire table, returns <see langword="null"/>.</remarks>
+        public string PartitionKey { get; }
+
+        /// <summary>When binding to a table entity, gets the row key of the entity.</summary>
+        /// <remarks>When binding to an entire table, returns <see langword="null"/>.</remarks>
+        public string RowKey { get; }
```
Comment (kshyju) [inline extensions/Worker.Extensions.Storage.Tables/src/TableInputAttribute.cs:47]: RowKey prop can be nullable.

**Review by kshyju — APPROVED** [review]

**Pull request merged**